ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRelFieldTrimmer.java

```diff
@@ -315,48 +322,128 @@ private boolean isRexLiteral(final RexNode rexNode) {
   }
 
 
+  private static class TableRefFinder extends RexVisitorImpl<Void> {
+    private Set<RexTableInputRef> tableRefs = null;
+    TableRefFinder() {
+      super(true);
+      this.tableRefs = new HashSet<>();
+    }
+
+    public Set<RexTableInputRef> getTableRefs() {
+      return this.tableRefs;
+    }
+
+    @Override
+    public Void visitTableInputRef(RexTableInputRef ref) {
+      this.tableRefs.add(ref);
+      return null;
+    }
+  }
+
   // Given a groupset this tries to find out if the cardinality of the grouping columns could have changed
   // because if not and it consist of keys (unique + not null OR pk), we can safely remove rest of the columns
   // if those are columns are not being used further up
   private ImmutableBitSet generateGroupSetIfCardinalitySame(final Aggregate aggregate,
                                         final ImmutableBitSet originalGroupSet, final ImmutableBitSet fieldsUsed) {
-    Pair<RelOptTable, List<Integer>> tabToOrgCol = HiveRelOptUtil.getColumnOriginSet(aggregate.getInput(),
-                                                                                     originalGroupSet);
-    if(tabToOrgCol == null) {
-      return originalGroupSet;
-    }
-    RelOptHiveTable tbl = (RelOptHiveTable)tabToOrgCol.left;
-    List<Integer> backtrackedGBList = tabToOrgCol.right;
-    ImmutableBitSet backtrackedGBSet = ImmutableBitSet.builder().addAll(backtrackedGBList).build();
 
-    List<ImmutableBitSet> allKeys = tbl.getNonNullableKeys();
-    ImmutableBitSet currentKey = null;
-    for(ImmutableBitSet key:allKeys) {
-      if(backtrackedGBSet.contains(key)) {
-        // only if grouping sets consist of keys
-        currentKey = key;
-        break;
+    RexBuilder rexBuilder = aggregate.getCluster().getRexBuilder();
+    RelMetadataQuery mq = aggregate.getCluster().getMetadataQuery();
+
+    Iterator<Integer> iterator = originalGroupSet.iterator();
+    Map<Pair<RelOptTable, Integer>, Pair<List<Integer>, List<Integer>>> mapGBKeysLineage= new HashMap<>();
```

Comment:
```List<Pair<Integer,Integer>>``` seems more intuitive/readable than holding two lists in the value?

Reply:
The reason this has to be kept this way is because later for a given table we need a list of all the back tracked keys and list of all the corresponding gb keys. If we keep List of pairs we will have to later create lists out of that.

Reply:
Where do you need additional lists? When you loop in L424, you are accessing both values. And in L408 you could just use loop (or a lambda) to add the left side from the pairs in the list. Or am I missing anything?

Reply:
@jcamachor 

> And in L408 you could just use loop (or a lambda would be nicer) to add the left side from the pairs in the list
But that would add an extra loop no?

Reply:
It's just a lambda vs the hit of instantiating two lists per pair (plus imo clearer code)... In any case, if you think it is not worth changing, that's OK.